File: src/main/kotlin/dev/ligature/test/LigatureSuite.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature.test

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.flow.toSet
import dev.ligature.Ligature
import dev.ligature.NamedElement

fun createSpec(creationFunction: () -> Ligature): StringSpec.() -> Unit {
    val testCollection = NamedElement("test")

    return {
        "Create and close store" {
            val store = creationFunction()
            store.read { tx ->
                tx.collections()
            }.toSet() shouldBe setOf()
            store.close()
        }

        "creating a new collection" {
            val store = creationFunction()
            store.write { tx ->
                tx.createCollection(testCollection)
            }
            store.read { tx ->
                tx.collections()
            }.toSet() shouldBe setOf(testCollection)
            store.close()
        }

        "access and delete new collection" {
            val store = creationFunction()
            store.write { tx ->
                tx.createCollection(testCollection)
            }
            store.read { tx ->
                tx.collections()
            }.toSet() shouldBe setOf(testCollection)
            store.write { tx ->
                tx.deleteCollection(testCollection)
                tx.deleteCollection(NamedElement("test2"))
            }
            store.read { tx ->
                tx.collections()
            }.toSet() shouldBe setOf()
            store.close()
        }

        "new collections should be empty" {
            val store = creationFunction()
            store.write { tx ->
                tx.createCollection(testCollection)
            }
            store.read { tx ->
                tx.allStatements(testCollection)
            }.toSet() shouldBe setOf()
            store.close()
        }

        "adding statements to collections" {
            val store = creationFunction()
            store.write { tx ->
                val ent1 = tx.newEntity(testCollection)
                val ent2 = tx.newEntity(testCollection)
                tx.addStatement(testCollection, Statement(ent1, a, ent2))
                tx.addStatement(testCollection, Statement(ent1, a, ent2))
            }
            store.read { tx ->
                tx.allStatements(testCollection)
            }.toSet() shouldBe
                    setOf(Statement(AnonymousEntity(1), a, AnonymousEntity(2)),
                           Statement(AnonymousEntity(1), a, AnonymousEntity(2)))
            store.close()
        }

        "removing statements from collections" {
            val store = creationFunction()
            store.write { tx ->
                val ent1 = tx.newEntity(testCollection)
                val ent2 = tx.newEntity(testCollection)
                val ent3 = tx.newEntity(testCollection)
                tx.addStatement(testCollection, Statement(ent1, a, ent2))
                tx.addStatement(testCollection, Statement(ent3, a, ent2))
                tx.removeStatement(testCollection, Statement(ent1, a, ent2))
            }
            store.read { tx ->
                tx.allStatements(testCollection)
            }.toSet() shouldBe
                    setOf(Statement(AnonymousEntity(3), a, AnonymousEntity(2)))
            store.close()
        }

        "new entity test" {
            val store = creationFunction()
            store.write { tx ->
                tx.addStatement(testCollection, Statement(tx.newEntity(testCollection), a, tx.newEntity(testCollection)))
                tx.addStatement(testCollection, Statement(tx.newEntity(testCollection), a, tx.newEntity(testCollection)))
            }
            store. compute { tx ->
                tx.allStatements(testCollection)
            }.toSet() shouldBe setOf(
                    Statement(AnonymousEntity(1), a, AnonymousEntity(2)),
                    Statement(AnonymousEntity(4), a, AnonymousEntity(5)))
            store.close()
        }

        "removing named entity" {
            val store = creationFunction()
            store.write { tx ->
                val ent1 = NamedEntity("a")
                val ent2 = NamedEntity("b")
                val ent3 = NamedEntity("c")
                tx.addStatement(testCollection, Statement(ent1, a, ent2))
                tx.addStatement(testCollection, Statement(ent3, a, ent2))
                tx.addStatement(testCollection, Statement(ent2, a, ent1))
                tx.removeEntity(testCollection, ent1)
            }
            store.read { tx ->
                tx.allStatements(testCollection)
            }.toSet() shouldBe
                    setOf(Statement(NamedEntity("c"), a, NamedEntity("b")))
            store.close()
        }

        "removing anonymous entity" {
            val store = creationFunction()
            store.write { tx ->
                val ent1 = tx.newEntity(testCollection)
                val ent2 = tx.newEntity(testCollection)
                val ent3 = tx.newEntity(testCollection)
                tx.addStatement(testCollection, Statement(ent1, a, ent2))
                tx.addStatement(testCollection, Statement(ent3, a, ent2))
                tx.addStatement(testCollection, Statement(ent2, a, ent1))
                tx.removeEntity(testCollection, ent1)
            }
            store.read { tx ->
                tx.allStatements(testCollection)
            }.toSet() shouldBe
                    setOf(Statement(AnonymousEntity(3), a, AnonymousEntity(2)))
            store.close()
        }

        "removing predicate" {
            val store = creationFunction()
            store.write { tx ->
                val ent1 = tx.newEntity(testCollection)
                val ent2 = tx.newEntity(testCollection)
                val ent3 = tx.newEntity(testCollection)
                tx.addStatement(testCollection, Statement(ent1, a, ent2))
                tx.addStatement(testCollection, Statement(ent3, Predicate("test"), ent2))
                tx.addStatement(testCollection, Statement(ent2, a, ent1))
                tx.removePredicate(testCollection, a)
            }
            store.read { tx ->
                tx.allStatements(testCollection)
            }.toSet() shouldBe
                    setOf(Statement(AnonymousEntity(3), Predicate("test"), AnonymousEntity(2)))
            store.close()
        }

        "matching against a non-existant collection" {
            val store = creationFunction()
            store.read { tx ->
                tx.matchStatements(testCollection, null, null, StringLiteral("French"))
                    .toSet() shouldBe setOf()
                tx.matchStatements(testCollection, null, a, null)
                    .toSet() shouldBe setOf()
            }
            store.close()
        }

        "matching statements in collections" {
            val store = creationFunction()
            lateinit var valjean: Entity
            lateinit var javert: Entity
            store.write { tx ->
                valjean = tx.newEntity(testCollection)
                javert = tx.newEntity(testCollection)
                tx.addStatement(testCollection, Statement(valjean, Predicate("nationality"), StringLiteral("French")))
                tx.addStatement(testCollection, Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601)))
                tx.addStatement(testCollection, Statement(javert, Predicate("nationality"), StringLiteral("French")))
            }
            store.read { tx ->
                tx.matchStatements(testCollection, null, null, StringLiteral("French"))
                        .toSet() shouldBe setOf(
                            Statement(valjean, Predicate("nationality"), StringLiteral("French")),
                            Statement(javert, Predicate("nationality"), StringLiteral("French"))
                )
                tx.matchStatements(testCollection, null, null, LongLiteral(24601))
                        .toSet() shouldBe setOf(
                            Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601))
                )
                tx.matchStatements(testCollection, valjean)
                        .toSet() shouldBe setOf(
                            Statement(valjean, Predicate("nationality"), StringLiteral("French")),
                            Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601))
                )
                tx.matchStatements(testCollection, javert, Predicate("nationality"), StringLiteral("French"))
                        .toSet() shouldBe setOf(
                            Statement(javert, Predicate("nationality"), StringLiteral("French"))
                )
                tx.matchStatements(testCollection, null, null, null)
                        .toSet() shouldBe setOf(
                            Statement(valjean, Predicate("nationality"), StringLiteral("French")),
                            Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601)),
                            Statement(javert, Predicate("nationality"), StringLiteral("French"))
                )
            }
            store.close()
        }

        "matching statements with literals and ranges in collections" {
            val store = creationFunction()
            lateinit var valjean: Entity
            lateinit var javert: Entity
            lateinit var trout: Entity
            store.write { tx ->
                valjean = tx.newEntity(testCollection)
                javert = tx.newEntity(testCollection)
                trout = tx.newEntity(testCollection)
                tx.addStatement(testCollection, Statement(valjean, Predicate("nationality"), StringLiteral("French")))
                tx.addStatement(testCollection, Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601)))
                tx.addStatement(testCollection, Statement(javert, Predicate("nationality"), StringLiteral("French")))
                tx.addStatement(testCollection, Statement(javert, Predicate("prisonNumber"), LongLiteral(24602)))
                tx.addStatement(testCollection, Statement(trout, Predicate("nationality"), StringLiteral("American")))
                tx.addStatement(testCollection, Statement(trout, Predicate("prisonNumber"), LongLiteral(24603)))
            }
            store.read { tx ->
                tx.matchStatements(testCollection, null, null, StringLiteralRange("French", "German"))
                        .toSet() shouldBe setOf(
                            Statement(valjean, Predicate("nationality"), StringLiteral("French")),
                            Statement(javert, Predicate("nationality"), StringLiteral("French"))
                )
                tx.matchStatements(testCollection, null, null, LongLiteralRange(24601, 24603))
                        .toSet() shouldBe setOf(
                            Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601)),
                            Statement(javert, Predicate("prisonNumber"), LongLiteral(24602))
                )
                tx.matchStatements(testCollection, valjean, null, LongLiteralRange(24601, 24603))
                        .toSet() shouldBe setOf(
                            Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601))
                )
            }
            store.close()
        }

//    "matching statements with collection literals in collections" {
//        val store = creationFunction()
//        val collection = store.createCollection(NamedEntity("test"))
//        collection shouldNotBe null
//        val tx = collection.writeTx()
//        TODO("Add values")
//        tx.commit()
//        val tx = collection.tx()
//        TODO("Add assertions")
//        tx.cancel() // TODO add test running against a non-existant collection w/ match-statement calls
//    }
    }
}

File: src/main/scala/dev/ligature/test/LigatureSuite.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature.test

import dev.ligature._
import dev.ligature.Ligature.a
import munit._

abstract class LigatureSuite extends FunSuite {
  def createLigature: Ligature

  val testDataset: Dataset = Dataset("test")

  test("Create and close store") {
    val res = createLigature.instance.use { instance  =>
      instance.read.use { tx =>
        tx.datasets.compile.toList
      }
    }.unsafeRunSync()
    assert(res.isEmpty)
  }

  test("creating a new dataset") {
    val res = createLigature.instance.use { instance =>
      for {
        _ <- instance.write.use { tx =>
          tx.createDataset(testDataset)
        }
        res <- instance.read.use { tx =>
          tx.datasets.compile.toList
        }
      } yield res
    }.unsafeRunSync().toSet
    assertEquals(res, Set(testDataset))
  }

  test("access and delete new dataset") {
    val res = createLigature.instance.use { instance  =>
      for {
        _ <- instance.write.use { tx =>
          tx.createDataset(testDataset)
        }
        _ <- instance.write.use { tx =>
          for {
            _ <- tx.deleteDataset(testDataset)
            _ <- tx.deleteDataset(Dataset("test2"))
          } yield ()
        }
        res <- instance.read.use { tx =>
           tx.datasets.compile.toList
         }
      } yield res
    }.unsafeRunSync()
    assert(res.isEmpty)
  }

  test("new datasets should be empty") {
    val res = createLigature.instance.use { instance  =>
      for {
        _ <- instance.write.use { tx =>
          tx.createDataset(testDataset)
        }
        res <- instance.read.use { tx =>
          tx.allStatements(testDataset).compile.toList
        }
      } yield res
    }.unsafeRunSync()
    assert(res.isEmpty)
  }

  test("new node test") {
    val res = createLigature.instance.use { instance  =>
      for {
        _ <- instance.write.use { tx =>
          for {
            nn1 <- tx.newNode(testDataset)
            nn2 <- tx.newNode(testDataset)
            _   <- tx.addStatement(testDataset, Statement(nn1, a, nn2))
            nn3 <- tx.newNode(testDataset)
            nn4 <- tx.newNode(testDataset)
            _   <- tx.addStatement(testDataset, Statement(nn3, a, nn4))
          } yield ()
        }
        res <- instance.read.use { tx =>
          tx.allStatements(testDataset).compile.toList
        }
      } yield res
    }.unsafeRunSync().toSet
    assertEquals(res.map { _.statement }, Set(
      Statement(AnonymousNode(1), a, AnonymousNode(2)),
      Statement(AnonymousNode(4), a, AnonymousNode(5))))
  }

  test("adding statements to datasets") {
    val res = createLigature.instance.use { instance  =>
      for {
        _ <- instance.write.use { tx =>
          for {
            ent1 <- tx.newNode(testDataset)
            ent2 <- tx.newNode(testDataset)
            _    <- tx.addStatement(testDataset, Statement(ent1, a, ent2))
            _    <- tx.addStatement(testDataset, Statement(ent1, a, ent2))
          } yield()
        }
        res  <- instance.read.use { tx =>
          tx.allStatements(testDataset).map { _.statement }.compile.toList
        }
      } yield res
    }.unsafeRunSync().toSet
    assertEquals(res, Set(Statement(AnonymousNode(1), a, AnonymousNode(2)),
      Statement(AnonymousNode(1), a, AnonymousNode(2))))
  }

  test("removing statements from datasets") {
    val res = createLigature.instance.use { instance =>
      for {
        _ <- instance.write.use { tx =>
          for {
            nn1 <- tx.newNode(testDataset)
            nn2 <- tx.newNode(testDataset)
            nn3 <- tx.newNode(testDataset)
            _ <- tx.addStatement(testDataset, Statement(nn1, a, nn2))
            _ <- tx.addStatement(testDataset, Statement(nn3, a, nn2))
            _ <- tx.removeStatement(testDataset, Statement(nn1, a, nn2))
            _ <- tx.removeStatement(testDataset, Statement(nn1, a, nn2))
            _ <- tx.removeStatement(testDataset, Statement(nn2, a, nn1))
          } yield ()
        }
        res <- instance.read.use { tx =>
          tx.allStatements(testDataset).map {
            _.statement
          }.compile.toList
        }
      } yield res
    }.unsafeRunSync().toSet
    assertEquals(res, Set(Statement(AnonymousNode(3), a, AnonymousNode(2))))
  }

////  test("matching against a non-existent dataset") {
////    val res = createLigature.instance.use { instance  =>
////    val (r1, r2) = instance.read.use { tx =>
////      for {
////        r1 <- tx.matchStatements(testDataset, null, null, StringLiteral("French")).compile.toList
////        r2 <- tx.matchStatements(testDataset, null, a, null).compile.toList
////      } yield(r1, r2)
////    }
////  }
////
////  test("matching statements in datasets") {
////    val res = createLigature.instance.use { instance  =>
////    lateinit var valjean: Node
////    lateinit var javert: Node
////    instance.write.use { tx =>
////      valjean = tx.newNode(testDataset)
////      javert = tx.newNode(testDataset)
////      tx.addStatement(testDataset, Statement(valjean, Predicate("nationality"), StringLiteral("French")))
////      tx.addStatement(testDataset, Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601)))
////      tx.addStatement(testDataset, Statement(javert, Predicate("nationality"), StringLiteral("French")))
////    }
////    instance.read.use { tx =>
////      tx.matchStatements(testDataset, null, null, StringLiteral("French"))
////              .toSet() shouldBe setOf(
////                  Statement(valjean, Predicate("nationality"), StringLiteral("French")),
////                  Statement(javert, Predicate("nationality"), StringLiteral("French"))
////      )
////      tx.matchStatements(testDataset, null, null, LongLiteral(24601))
////              .toSet() shouldBe setOf(
////                  Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601))
////      )
////      tx.matchStatements(testDataset, valjean)
////              .toSet() shouldBe setOf(
////                  Statement(valjean, Predicate("nationality"), StringLiteral("French")),
////                  Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601))
////      )
////      tx.matchStatements(testDataset, javert, Predicate("nationality"), StringLiteral("French"))
////              .toSet() shouldBe setOf(
////                  Statement(javert, Predicate("nationality"), StringLiteral("French"))
////      )
////      tx.matchStatements(testDataset, null, null, null)
////              .toSet() shouldBe setOf(
////                  Statement(valjean, Predicate("nationality"), StringLiteral("French")),
////                  Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601)),
////                  Statement(javert, Predicate("nationality"), StringLiteral("French"))
////      )
////    }
////  }
////
////  test("matching statements with literals and ranges in datasets") {
////    val res = createLigature.instance.use { instance  =>
////    lateinit var valjean: Node
////    lateinit var javert: Node
////    lateinit var trout: Node
////    instance.write.use { tx =>
////      valjean = tx.newNode(testDataset)
////      javert = tx.newNode(testDataset)
////      trout = tx.newNode(testDataset)
////      tx.addStatement(testDataset, Statement(valjean, Predicate("nationality"), StringLiteral("French")))
////      tx.addStatement(testDataset, Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601)))
////      tx.addStatement(testDataset, Statement(javert, Predicate("nationality"), StringLiteral("French")))
////      tx.addStatement(testDataset, Statement(javert, Predicate("prisonNumber"), LongLiteral(24602)))
////      tx.addStatement(testDataset, Statement(trout, Predicate("nationality"), StringLiteral("American")))
////      tx.addStatement(testDataset, Statement(trout, Predicate("prisonNumber"), LongLiteral(24603)))
////    }
////    instance.read.use { tx =>
////      tx.matchStatements(testDataset, null, null, StringLiteralRange("French", "German"))
////              .toSet() shouldBe setOf(
////                  Statement(valjean, Predicate("nationality"), StringLiteral("French")),
////                  Statement(javert, Predicate("nationality"), StringLiteral("French"))
////      )
////      tx.matchStatements(testDataset, null, null, LongLiteralRange(24601, 24603))
////              .toSet() shouldBe setOf(
////                  Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601)),
////                  Statement(javert, Predicate("prisonNumber"), LongLiteral(24602))
////      )
////      tx.matchStatements(testDataset, valjean, null, LongLiteralRange(24601, 24603))
////              .toSet() shouldBe setOf(
////                  Statement(valjean, Predicate("prisonNumber"), LongLiteral(24601))
////      )
////    }
////  }
////
////  test("matching statements with dataset literals in datasets") {
////    val res = createLigature.instance.use { instance  =>
////    val dataset = store.createDataset(NamedNode("test"))
////    dataset shouldNotBe null
////    val tx = dataset.writeTx()
////    TODO("Add values")
////    tx.commit()
////    val tx = dataset.tx()
////    TODO("Add assertions")
////    tx.cancel() // TODO add test running against a non-existant dataset w/ match-statement calls
////  }
}
